.github/workflows/build-and-test.yml

```diff
@@ -2,48 +2,77 @@ name: build-and-test
 
 on:
   push:
-    branches: [ main ]
-    paths:
-    - '**.cs'
-    - '**.csproj'
+  pull_request:
   workflow_dispatch:
     inputs:
       reason:
         description: 'The reason for running a build?'
         required: true
         default: 'Manual build'
 
-env:
-  DOTNET_VERSION: '8.0.x' # set this to the dot net version to use
-
 jobs:
   build:
     name: build-${{matrix.os}}
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-latest, windows-latest, macOS-latest]
+        os: [ ubuntu-latest, windows-latest, macos-latest ]
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          filter: tree:0
+
+      - name: 'Print manual build reason'
+        if: ${{ github.event_name == 'workflow_dispatch' }}
+        run: |
+          echo 'Reason: ${{ github.event.inputs.reason }}'
+
+      - name: Setup .NET
+        uses: actions/setup-dotnet@v4
+        with:
+          global-json-file: global.json
+
+      - name: Install dependencies
+        run: |
+          dotnet restore
+
+      - name: Build
+        run: |
+          dotnet build --configuration Release --no-restore
+
+      - uses: actions/upload-artifact@v4
+        if: matrix.os == 'ubuntu-latest'
+        with:
+          name: nuget
+          path: artifacts/package/release/*.nupkg
+
+      - name: Test
+        run: dotnet test --configuration Release --no-restore
+
+      - name: Publish
+        if: matrix.os == 'ubuntu-latest' && github.ref_type == 'tag'
+        run: |
+          dotnet nuget push --skip-duplicate artifacts/package/release/*.nupkg \
+            --source "https://api.nuget.org/v3/index.json" --api-key "${{ secrets.NUGET_API_KEY }}"
+
+  test-samples:
```

Comment:
> But, I really don't want these to run with every push and pull_request

Why not? Open source github actions are free and it runs pretty quick so imo it's better to know if you've broken them in any way right away.

Reply:
How long will these take to run? I've seen it take up to two full minutes for building a single container, and running it. This is proposing that we'd have two built, with imaginably more for each test project sample we add. It's not even happening in parallel, so each time this runs we could be seeing four minute runs on a PR, that's a long time IMO (basically add 2 minutes for every project sample 🙈). We'll write more unit tests.

Reply:
Currently, it takes [46s total](https://github.com/IEvangelist/dotnet-github-actions-sdk/actions/runs/8051843926/job/21999555147?pr=6) to build and run 2 samples, and it might be even faster with `Microsoft.NET.Build.Containers`. It runs in parallel with normal build, so it literally doesn't matter.

> We'll write more unit tests.

Will unit tests be able to catch changes made by github?

Reply:
Currently, it takes [46s total](https://github.com/IEvangelist/dotnet-github-actions-sdk/actions/runs/8051843926/job/21999555147?pr=6) to build and run 2 samples, and it might be even faster with `Microsoft.NET.Build.Containers`. It runs in parallel with normal build, so it literally doesn't matter.

> We'll write more unit tests.

Will unit tests be able to catch changes made by github?

Reply:
I suppose that's not the end of the world, but I reserve the right to pull this out if I get annoyed with it later. 😛